packages/tizen-remote/lib/index.js

```diff
@@ -0,0 +1,742 @@
+    }
+    const send = /** @type {(data: any) => Promise<void>} */ (
+      promisify(/** @type {WebSocket} */ (ws).send).bind(this.#ws)
+    );
+    await send(payload);
+  }
+
+  /**
+   * Send JSON-serializable data to the Tizen web socket server.
+   *
+   * This is low-level, and you probably want something else.
+   * @template T
+   * @param {string} channel
+   * @param {any} data
+   * @param {SendOptions} [opts]
+   * @returns {Promise<T>}
+   */
+  async sendRequest(channel, data, {noConnect = false} = {}) {
+    /** @type {string} */
+    let payload;
+    /** @type {WebSocket} */
+    let ws;
+    if (this.#ws) {
+      ws = this.#ws;
+    } else {
+      if (noConnect) {
+        throw new Error(`Disconnected; cannot send message: ${data}`);
+      }
+      this.#debug('Not connected; attempting connection...');
+      ws = await this.connect();
+    }
+
+    try {
+      payload = JSON.stringify(data);
+    } catch (err) {
+      throw new Error(`Unable to serialize data to JSON: ${/** @type {Error} */ (err).message}`);
+    }
+
+    const send = /** @type {(data: any) => Promise<void>} */ (promisify(ws.send).bind(this.#ws));
+
+    this.#debug('Sending request on channel %s: %O', channel, data);
+    await send(payload);
+
+    return await new Promise((resolve, reject) => {
+      const tokenTimer = setTimeout(() => {
+        reject(new Error(`Did not receive token in ${this.#tokenTimeout}ms`));
+      }, this.#tokenTimeout);
+
+      /** @param {import('ws').RawData} data */
+      const listener = (data) => {
+        try {
+          const resData = JSON.parse(data.toString());
+          if (resData.event === channel) {
+            resolve(resData);
+          }
+        } catch {
+          // if we can't parse the data, it's not something we're interested in
+        } finally {
+          clearTimeout(tokenTimer);
+          this.#offWs(WsEvent.MESSAGE, listener);
+        }
+      };
+      this.#onWs(WsEvent.MESSAGE, listener);
+    });
+  }
+
+  /**
+   * @template {(...args: any[]) => void} Listener
+   * @param {WsEvent} event
+   * @param {Listener} listener
+   * @param {{ context?: any }} [opts]
+   * @returns {Listener}
+   */
+  #onWs(event, listener, {context} = {}) {
+    if (this.#ws) {
+      const listeners = this.#listeners.get(event) ?? new Set();
+      this.#listeners.set(event, listeners);
+      if (context) {
+        const boundListener = /** @type {Listener} */(listener.bind(context));
+        listeners.add(boundListener);
+        this.#ws.on(event, boundListener);
+        return boundListener;
+      }
+      listeners.add(listener);
+      this.#ws.on(event, listener);
+      return listener;
+    }
+    throw new Error('Not connected');
+  }
+
+  /**
+   *
+   * @template {(...args: any[]) => void} Listener
+   * @param {WsEvent} event
+   * @param {Listener} listener
+   * @param {{ context?: any }} [opts]
+   * @returns {Listener}
+   */
+  #onceWs(event, listener, {context} = {}) {
```

Comment:
this looks like pretty much an exact duplicate of `#onWs` with a different listener method (`once` vs `on`) suggesting these can be consolidated